test/dynamo/test_autograd_function.py

```diff
@@ -468,6 +468,62 @@ def fn(a, b):
         self.assertEqual(res, MyMM.apply(a, a))
         self.assertEqual(cnt.frame_count, 1)
 
+    def test_set_materialize_grads(self):
+        cnt = torch._dynamo.testing.CompileCounterWithBackend("aot_eager")
+
+        class MulY(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                ctx.set_materialize_grads(False)
```

Comment:
I thought about this some more. The behavior before this PR (where we ignore set_materialize_grads) is actually correct. Here's why:
- At Dynamo time, we trace the forward and the backward of the autograd.Function
- when we trace the backward, we assume that all of the gradients are Tensors. So we've baked in the assumption that our gradients are materialized
- it is actually wrong for us to reconstruct the autograd.Function with set_materialize_grads = False.

Given that, to close this issue, we should construct a test case that demonstrates this (somehow get a situation where the backward receives a None gradient) and then test an autograd.Function with ctx.set_materialize_grads(False) and ship just the test case.